### PR TITLE
basic async addition for audio renderer (not for all parts yet)

### PR DESCRIPTION
**Pull request by 16-Bit-Dog**
A mix of async thread work and keeping dedicated thread time to a non-async process to reduce latency of a task: slight reduction in time to complete audio rendering

This was a small addition; more complex changes are going to be done later...

### CODE REVIEW

**Comment by 16-Bit-Dog** [comment]
oof, gotta fix that format of code to follow clang

File: src/audio_core/audio_renderer.cpp [hunk 1]
```diff
@@ -68,6 +69,11 @@ namespace {
 } // namespace
 
 namespace AudioCore {
+
+std::vector<std::future<void>> queueMixedThreadFence;
```
Comment (lioncash) [inline src/audio_core/audio_renderer.cpp:73]: We don't use global variables

File: src/audio_core/audio_renderer.cpp [hunk 2]
```diff
@@ -236,6 +248,7 @@ void AudioRenderer::QueueMixedBuffer(Buffer::Tag tag) {
             mix_buffers[i] =
                 command_generator.GetMixBuffer(in_params.buffer_offset + buffer_offsets[i]);
         }
+        queueMixedThreadFence[queueMixedThreadFence.size() - 1].get();
```
Comment (lioncash) [inline src/audio_core/audio_renderer.cpp:251]:
```suggestion
        queueMixedThreadFence.back().get();
```

File: src/audio_core/audio_renderer.cpp [hunk 3]
```diff
@@ -313,11 +326,16 @@ void AudioRenderer::QueueMixedBuffer(Buffer::Tag tag) {
     elapsed_frame_count++;
     voice_context.UpdateStateByDspShared();
 }
-
 void AudioRenderer::ReleaseAndQueueBuffers() {
+
     const auto released_buffers{audio_out->GetTagsAndReleaseBuffers(stream)};
+
+    queueMixedThreadFence.resize(0); // instead of passing a s16 to the queue mixed buffer to
```
Comment (lioncash) [inline src/audio_core/audio_renderer.cpp:333]: Comments go on top of code rather than beside it

File: src/audio_core/audio_renderer.cpp [hunk 4]
```diff
@@ -313,11 +326,16 @@ void AudioRenderer::QueueMixedBuffer(Buffer::Tag tag) {
     elapsed_frame_count++;
     voice_context.UpdateStateByDspShared();
 }
-
 void AudioRenderer::ReleaseAndQueueBuffers() {
+
```
Comment (lioncash) [inline src/audio_core/audio_renderer.cpp:330]: Unnecessary newline

File: src/audio_core/audio_renderer.cpp [hunk 5]
```diff
@@ -313,11 +326,16 @@ void AudioRenderer::QueueMixedBuffer(Buffer::Tag tag) {
     elapsed_frame_count++;
     voice_context.UpdateStateByDspShared();
 }
-
```
Comment (lioncash) [inline src/audio_core/audio_renderer.cpp:316]: Newline should be left in here

File: src/audio_core/audio_renderer.cpp [hunk 6]
```diff
@@ -195,7 +201,9 @@ ResultCode AudioRenderer::UpdateAudioRenderer(const std::vector<u8>& input_param
         return AudioCommon::Audren::ERR_INVALID_PARAMETERS;
     }
 
-    ReleaseAndQueueBuffers();
+    keepThreadReady1 = std::async(std::launch::async, [&] { ReleaseAndQueueBuffers(); });
+
+    keepThreadReady1.get();
```
Comment (lioncash) [inline src/audio_core/audio_renderer.cpp:206]: What is the purpose of this async if it's being created and then immediately being blocked on?

File: src/audio_core/audio_renderer.cpp [hunk 7]
```diff
@@ -313,11 +326,16 @@ void AudioRenderer::QueueMixedBuffer(Buffer::Tag tag) {
...
+
     const auto released_buffers{audio_out->GetTagsAndReleaseBuffers(stream)};
+
+    queueMixedThreadFence.resize(0); // instead of passing a s16 to the queue mixed buffer to
+                                     // control, pushing values to a vector seemed about as fast
+
     for (const auto& tag : released_buffers) {
-        QueueMixedBuffer(tag);
+        keepThreadReady2 = std::async(std::launch::async, [&] { QueueMixedBuffer(tag); });
+        keepThreadReady2.get();
```
Comment (lioncash) [inline src/audio_core/audio_renderer.cpp:338]: ditto

**Comment by ogniK5377** [comment]
I would suggest setting the async audio to be more closer to a mailbox type of implementation too much functionality of the switches fw

**Comment by 16-Bit-Dog** [comment]
thank you for all of the recommendations and pointers; I'll close the PR now to fix it up and implement more (have an idea for a interesting barrier-mailbox system I already thought up which I think may work well)

**Pull request closed without merging**